features/mbedtls/targets/TARGET_CRYPTOCELL310/ecdsa_alt.c

```diff
@@ -0,0 +1,325 @@
+/*
+ *  ecdsa_alt.c
+ *
+ *  Copyright (C) 2018, ARM Limited, All Rights Reserved
+ *  SPDX-License-Identifier: Apache-2.0
+ *
+ *  Licensed under the Apache License, Version 2.0 (the "License"); you may
+ *  not use this file except in compliance with the License.
+ *  You may obtain a copy of the License at
+ *
+ *  http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *  Unless required by applicable law or agreed to in writing, software
+ *  distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ *  WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *  See the License for the specific language governing permissions and
+ *  limitations under the License.
+ *
+ */
+
+#include "mbedtls/ecdsa.h"
+#include <string.h>
+#include "crys_ecpki_ecdsa.h"
+#include "crys_ecpki_build.h"
+#include "crys_common.h"
+#include "crys_ecpki_kg.h"
+#include "crys_ecpki_domain.h"
+#include "crys_ec_edw_api.h"
+#include "mbedtls/platform.h"
+#include "cc_internal.h"
+
+/* Implementation that should never be optimized out by the compiler */
+static void mbedtls_zeroize( void *v, size_t n ) {
+    volatile unsigned char *p = (unsigned char*)v;
+    while( n-- ) *p++ = 0;
+}
+
+static CRYS_ECPKI_HASH_OpMode_t message_size_to_hash_mode( size_t blen )
+{
+    CRYS_ECPKI_HASH_OpMode_t hash_mode;
+    switch( blen )
+    {
+    case CRYS_HASH_SHA1_DIGEST_SIZE_IN_WORDS*sizeof(uint32_t):
+            hash_mode = CRYS_ECPKI_AFTER_HASH_SHA1_mode;
+    break;
+    case CRYS_HASH_SHA224_DIGEST_SIZE_IN_WORDS*sizeof(uint32_t):
+            hash_mode = CRYS_ECPKI_AFTER_HASH_SHA224_mode;
+    break;
+    case CRYS_HASH_SHA256_DIGEST_SIZE_IN_WORDS*sizeof(uint32_t):
+            hash_mode = CRYS_ECPKI_AFTER_HASH_SHA256_mode;
+    break;
+    case CRYS_HASH_SHA384_DIGEST_SIZE_IN_WORDS*sizeof(uint32_t):
+            hash_mode = CRYS_ECPKI_AFTER_HASH_SHA384_mode;
+    break;
+    case CRYS_HASH_SHA512_DIGEST_SIZE_IN_WORDS*sizeof(uint32_t):
+            hash_mode = CRYS_ECPKI_AFTER_HASH_SHA512_mode;
+    break;
+    default:
+        hash_mode = CRYS_ECPKI_HASH_OpModeLast;
+    }
+
+    return hash_mode;
+}
+
+#if defined(MBEDTLS_ECDSA_SIGN_ALT)
+int mbedtls_ecdsa_sign( mbedtls_ecp_group *grp, mbedtls_mpi *r, mbedtls_mpi *s,
+                const mbedtls_mpi *d, const unsigned char *buf, size_t blen,
+                int (*f_rng)(void *, unsigned char *, size_t), void *p_rng )
+{
+    int ret = 0;
+    CRYSError_t CrysRet = CRYS_OK;
+    void* pHeap = NULL;
+    size_t heapSize = 0;
+    uint8_t* pSignature = NULL;
+    CRYS_ECPKI_HASH_OpMode_t hash_mode = message_size_to_hash_mode( blen );
+    uint32_t signature_size =  ( ( grp->nbits + 7 ) / 8 ) *2;
+    const uint32_t signature_size_for_heap = signature_size;
+    mbedtls_rand_func_container cc_rand = { f_rng, p_rng };
+    const CRYS_ECPKI_Domain_t*  pDomain =  CRYS_ECPKI_GetEcDomain ( convert_mbedtls_grp_id_to_crys_domain_id( grp->id ) );
```

Comment:
I would suggest moving the assignment to until after the parameter checks.

Reply:
It's more efficient to initialize a variable with a  value, rather than assign it later.
Nonetheless, I don't think this is a blocker